File: LeetCode/1-100/98. 验证二叉搜索树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
  let temp = 0
  let flag = true
  let start = false
  const Ftree = (root) => {
    if (!flag) return
    if (root.left) Ftree(root.left)
    if (root.val <= temp && start) {
      flag = false
    } else {
      temp = root.val
    }
    start = true
    if (root.right) Ftree(root.right)
  }
  Ftree(root)
  return flag
}

File: LeetCode/1-100/98. 验证二叉搜索树.go

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func isValidBST(root *TreeNode) bool {
	return find(root, math.MinInt, math.MaxInt)
}
func find(root *TreeNode, min int, max int) bool {
	if root == nil {
		return true
	}
	if root.Val <= min || root.Val >= max {
		return false
	}
	return find(root.Left, min,root.Val) && find(root.Right, root.Val, max)
}


 func isValidBST(root *TreeNode) bool {
	var find func(*TreeNode, int, int) bool
		find = func(node *TreeNode, min int, max int) bool {
			if node == nil {
				return true
			}
			if node.Val <= min || node.Val >= max {
				return false
			}
			return find(node.Left, min, node.Val) && find(node.Right, node.Val, max)
		}
		return find(root, math.MinInt, math.MaxInt)
	}